tests/test_anomaly_functionsv2.py

```diff
@@ -0,0 +1,159 @@
+from iotfunctions.enginelog import EngineLogging
+from iotfunctions.anomaly import SaliencybasedGeneralizedAnomalyScoreV2, SpectralAnomalyScoreExt, \
+                                 FFTbasedGeneralizedAnomalyScoreV2, KMeansAnomalyScoreV2
+from nose.tools import assert_true, nottest
+
+# constants
+Temperature = 'Temperature'
+kmeans = 'TemperatureKmeansScore'
+fft = 'TemperatureFFTScore'
+spectral = 'TemperatureSpectralScore'
+spectralinv = 'TemperatureSpectralScoreInv'
+sal = 'SaliencyAnomalyScore'
+gen = 'TemperatureGeneralizedScore'
+
+logger = logging.getLogger('Test Regressor')
+
+@nottest
+class DatabaseDummy:
+    tenant_id = '###_IBM_###'
+    db_type = 'db2'
+    model_store = FileModelStore('/tmp')
+    def _init(self):
+        return
+
+
+def test_anomaly_scores():
+
+    numba_logger = logging.getLogger('numba')
+    numba_logger.setLevel(logging.ERROR)
+
+
+    #####
+    print('Create dummy database')
+    db_schema=None
+    db = DatabaseDummy()
+    print (db.model_store)
+
+    #####
+
+    jobsettings = { 'db': db, '_db_schema': 'public'}
+    EngineLogging.configure_console_logging(logging.DEBUG)
+
+    # Run on the good pump first
+    # Get stuff in
+    print('Read Anomaly Sample data in')
+    df_i = pd.read_csv('./data/AzureAnomalysample.csv', index_col=False, parse_dates=['timestamp'])
+
+    df_i['entity'] = 'MyRoom'
+    df_i[Temperature] = df_i['value'] + 20
+    df_i = df_i.drop(columns=['value'])
+
+    # and sort it by timestamp
+    df_i = df_i.sort_values(by='timestamp')
+    df_i = df_i.set_index(['entity', 'timestamp']).dropna()
+
+    for i in range(0, df_i.index.nlevels):
+        print(str(df_i.index.get_level_values(i)))
+
+    #####
+
+    print('Compute Spectral Anomaly Score')
+    spsi = SpectralAnomalyScoreExt(Temperature, 12, spectral, spectralinv)
+    et = spsi._build_entity_type(columns=[Column(Temperature, Float())], **jobsettings)
+    spsi._entity_type = et
+    df_i = spsi.execute(df=df_i)
+
+    print('Compute Saliency Anomaly Score')
+    sali = SaliencybasedGeneralizedAnomalyScoreV2(Temperature, 12, True, sal)
+    et = sali._build_entity_type(columns=[Column(Temperature, Float())], **jobsettings)
+    sali._entity_type = et
+    df_i = sali.execute(df=df_i)
+
+    print('Compute FFT Anomaly Score')
+    ffti = FFTbasedGeneralizedAnomalyScoreV2(Temperature, 12, True, fft)
+    et = ffti._build_entity_type(columns=[Column(Temperature, Float())], **jobsettings)
+    ffti._entity_type = et
+    df_i = ffti.execute(df=df_i)
+
+    print('Compute K-Means Anomaly Score')
+    kmi = KMeansAnomalyScoreV2(Temperature, 12, True, kmeans)
+    et = kmi._build_entity_type(columns=[Column(Temperature, Float())], **jobsettings)
+    kmi._entity_type = et
+    df_comp = kmi.execute(df=df_i)
+
+    print("Executed Anomaly functions")
+
+    # df_comp.to_csv('./data/AzureAnomalysampleOutputV2.csv')
+    df_o = pd.read_csv('./data/AzureAnomalysampleOutputV2.csv')
```

Comment:
How are we generating these ground truth values?

Reply:
The tell-tale 'to_csv()' is backed up by a visual inspection, so while I'm technically cheating I generated the values originally from the MonitorAnomalyFunctions notebook.
Since the purpose of anomaly scorers is classification, I should have classified the raw data and test whether the scorers find the anomalies as expected. However, there are subtle differences depending on method and alert trigger (aka decision boundary).
![image](https://user-images.githubusercontent.com/11428923/108484651-56b37c00-729c-11eb-81bf-782d3d95dffe.png)
Furthermore I'm quite interested to detect major behavior changes, for example a score that detects anomalies as 'reliable' as before, but with a different decision boundary, because an underlying library has been upgraded. So I'm simply comparing the old and known scores representing known and acceptable behavior with new and potentially undesired behavior.
It's a guard rail to protect us from falling, nothing more.